### PR TITLE
DT: new pinctrl helpers

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
The commit which introduces new pin control helper macros to devicetree.h is:

```
devicetree: add devicetree/pinctrl.h

This contains accessor macros for getting phandles out of pinctrl
properties by name and index. As usual, the representation in C for a
phandle is a node identifier.

Add these new macros:

- DT_PINCTRL_BY_IDX(node_id, pc_idx, idx): phandle at index idx
  in the pinctrl-<pc_idx> property

- DT_PINCTRL_0(node_id, idx): pinctrl-0 convenience for the same

- DT_PINCTRL_BY_NAME(node_id, name, idx): phandle at index idx
  in the pinctrl property named 'name'

- DT_PINCTRL_NAME_TO_IDX(node_id, name): convert a pinctrl property
  name to its index number

- DT_NUM_PINCTRLS_BY_IDX(node_id, pc_idx): number of phandles in
  pinctrl-<pc_idx>

- DT_NUM_PINCTRLS_BY_NAME(node_id, name): number of phandles in a
  named pinctrl property

- DT_NUM_PINCTRL_STATES(node_id): total number of pinctrl-<pc_idx>
  properties

- DT_PINCTRL_HAS_IDX(node_id, pc_idx): does pinctrl-<pc_idx> exist?

- DT_PINCTRL_HAS_NAME(node_id, name): does a named pinctrl property
  exist?

- DT_PINCTRL_IDX_TO_NAME_TOKEN(node_id, pc_idx): convert a pinctrl
  index to its name as a token, similar to DT_STRING_TOKEN()

- DT_PINCTRL_IDX_TO_NAME_UPPER_TOKEN(node_id, pc_idx): like
  DT_PINCTRL_IDX_TO_NAME_TOKEN, but with an uppercase result

As well as DT_DRV_INST equivalents, which take inst wherever node_id
appears above:

- DT_INST_PINCTRL_BY_IDX()
- DT_INST_PINCTRL_0()
- DT_INST_PINCTRL_BY_NAME()
- DT_INST_PINCTRL_NAME_TO_IDX()
- DT_INST_NUM_PINCTRLS_BY_IDX()
- DT_INST_NUM_PINCTRLS_BY_NAME()
- DT_INST_NUM_PINCTRL_STATES()
- DT_INST_PINCTRL_HAS_IDX()
- DT_INST_PINCTRL_HAS_NAME()
- DT_INST_PINCTRL_IDX_TO_NAME_TOKEN()
- DT_INST_PINCTRL_IDX_TO_NAME_UPPER_TOKEN()


Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>
```

The rest of the PR is prep work, bookkeeping, and documentation.

---
This change is [Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/37644)

### CODE REVIEW

File: include/devicetree/pinctrl.h
Comment (gmarull) [inline include/devicetree/pinctrl.h:72]: I'm not sure we should add this one. I think that `pinctrl` is better represented using states, so drivers should work using the state concept and not indices, so I'm not sure if providing extra helpers for indices helps on that front.

This code:
```
      n: node {
              pinctrl-0 = <&foo &bar>;
              pinctrl-1 = <&baz &blub>;
              pinctrl-names = "default", "sleep";
      }
 ```

and this other one:
```
      n: node {
              pinctrl-0 = <&foo &bar>;
              pinctrl-1 = <&baz &blub>;
              pinctrl-names = "sleep", "default";
      }
 ```

should result in same behavior, ie indices are not relevant.

File: dts/bindings/test/vnd,adc-temp-sensor.yaml
Comment (gmarull) [inline dts/bindings/test/vnd,adc-temp-sensor.yaml:29]: should we provide something like this: https://github.com/zephyrproject-rtos/zephyr/blob/ca4c7d200ee6f1ef77dbfe1ffef8be537af140da/dts/bindings/pinctrl/pinctrl-device.yaml ?

**Review by gmarull — COMMENTED** [review]
Some initial thoughts, thanks for this DT update

File: include/devicetree/pinctrl.h
Comment (mbolivar-nordic) [reply include/devicetree/pinctrl.h:72]: > so drivers should work using the state concept and not indices

This seems to be the main assertion here. The fact is that drivers are currently using indices, not just states. Now, that may be because there is no support for names yet. I don't know.

Does Linux use states exclusively, or does it also use indices?

Comment (gmarull) [reply include/devicetree/pinctrl.h:72]: Yes, Linux uses states: https://www.kernel.org/doc/html/latest/driver-api/pin-control.html#pin-control-requests-from-drivers

As far as I know Zephyr has never had such concept (probably because DT API/tooling was missing and/or current pinctrl drivers are not in a mature status).

Comment (galak) [reply include/devicetree/pinctrl.h:72]: I see this as no different than any other `name`d property in devicetree.  Its left to the binding and driver to decide if it wants to use the named version or not, and if the order of the values in an "array" has meaning.  For example regs and irqs do this all the time with assumptions in binding/driver.

File: dts/bindings/test/vnd,adc-temp-sensor.yaml
Comment (galak) [reply dts/bindings/test/vnd,adc-temp-sensor.yaml:29]: @mbolivar-nordic I don't know if you want to add `pinctrl-0` & `pinctrl-1`, and `pinctrl-names` to dts/bindings/base/base.yaml as part of this PR.

**Review by galak — CHANGES_REQUESTED** [review]
Looks good, one comment on what `_BY_IDX` means and impact on `PINCTRL_0` api variants.

File: include/devicetree/pinctrl.h
Comment (gmarull) [reply include/devicetree/pinctrl.h:72]: Whether it's the driver or not who takes care of storing states, names makes things easier. If every driver decides to use indices on their own, bindings will need to specify "use index 0 for default state and 1 for sleep".

If I have to write a driver that needs to obtain a particular IRQ (e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/st/h7/stm32h7.dtsi#L452) I'll always obtain it by name, just to make sure I get what I want regardless of the order.

Comment (galak) [reply include/devicetree/pinctrl.h:72]: > Whether it's the driver or not who takes care of storing states, names makes things easier. If every driver decides to use indices on their own, bindings will need to specify "use index 0 for default state and 1 for sleep".
> 
> If I have to write a driver that needs to obtain a particular IRQ (e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/st/h7/stm32h7.dtsi#L452) I'll always obtain it by name, just to make sure I get what I want regardless of the order.

I agree that is better practice, but from an API point of view it makes sense to have what @mbolivar-nordic proposes in this PR.  How that gets used is left the the driver developers.

Comment (gmarull) [reply include/devicetree/pinctrl.h:72]: > I agree that is better practice, but from an API point of view it makes sense to have what @mbolivar-nordic proposes in this PR. How that gets used is left the the driver developers.

If we think pinctrl representation is better understood using states, as in Linux, then indices should have no relevance. We now use index 0 everywhere because it's really the only thing we have, but as soon as more states start to be implemented then `0` will have no special meaning other than being the first entry. Summarizing, I think we should be careful on adding new APIs, specially because current pinctrl state is not mature enough. It may be useful as a transition step to make current code less verbose, though.

Comment (mbolivar-nordic) [reply include/devicetree/pinctrl.h:72]: Trying to summarize:

- @gmarull wants Zephyr to use states exclusively, thinks using indices is a bad practice, points at Linux pinmux APIs using states
- @galak agrees using names is better practice, but thinks this should be left up to the user to decide and the API should be neutral

I don't know what the project is going to decide re: states vs. indices.

However, if we do decide on a state-based approach, then doing all the index-based pin control access via dedicated APIs instead of directly accessing the pinctrl-0 and pinctrl-1 properties as we are doing now will make it trivial to deprecate the index-based way of doing things. So even if the API is not forever, I think it is better to introduce and use it than let the current status quo stand.

@gmarull would you agree with that argument?

File: dts/bindings/test/vnd,adc-temp-sensor.yaml
Comment (mbolivar-nordic) [reply dts/bindings/test/vnd,adc-temp-sensor.yaml:29]: I think I'll just include pinctrl-device.yaml into vnd,adc-temp-sensor for now. I think base.yaml is pretty big already and it's hard for users to know when a property in there is relevant to their device or not.

Comment (mbolivar-nordic) [reply dts/bindings/test/vnd,adc-temp-sensor.yaml:29]: @gmarull actually, that commit doesn't seem to be upstream yet...

Comment (gmarull) [reply dts/bindings/test/vnd,adc-temp-sensor.yaml:29]: It's part of #37572, feel free to take that file.

**Comment by mbolivar-nordic** [comment]
@galak @gmarull please revisit. I updated the PR to hopefully address all the use cases and comments I could. I replied to comments individually when I didn't make any change.

File: soc/arm/nuvoton_npcx/common/soc_dt.h
Comment (ChiHuaL) [inline soc/arm/nuvoton_npcx/common/soc_dt.h:166]: Because the parameter is "inst", it should use DT_DRV_INST equivalent macro here.
`
UTIL_LISTIFY(DT_INST_NUM_PINCTRLS_BY_IDX(inst, 0),      \
`

File: include/devicetree/pinctrl.h
Comment (galak) [inline include/devicetree/pinctrl.h:129]: Why is this not just `DT_PROP_LEN` of the property?  Not sure we need this macro.

File: include/devicetree/pinctrl.h
Comment (galak) [inline include/devicetree/pinctrl.h:153]: same comment as above.

**Review by galak — CHANGES_REQUESTED** [review]
looks good, just question about need of DT_NUM_PINCTRLS_BY_IDX and variants.

File: include/devicetree/pinctrl.h
Comment (mbolivar-nordic) [reply include/devicetree/pinctrl.h:129]: > Why is this not just `DT_PROP_LEN` of the property? Not sure we need this macro.

I explained a bit here:

https://github.com/zephyrproject-rtos/zephyr/pull/37621#discussion_r687987955

That's what people are doing now, but they are pasting ``pinctrl_##num`` in their code to do that. Pasting in cpp has "gotchas" that the DT_CATx macros we use under the hood make sure users don't run into.

For example if you do

```c
#define MY_INDEX 1
#define FOO(node_id, pc_idx) DT_PROP_LEN(node_id, pinctrl_ ## pc_idx)

void foo(void)
{
	printk("raw len with pasting: %d\n", FOO(TEST_TEMP, MY_INDEX));
	printk("len with api: %d\n", DT_NUM_PINCTRLS_BY_IDX(TEST_TEMP, MY_INDEX));
}
```

Then this expands to:

```c
void foo(void)
{
 printk("raw len with pasting: %d\n", DT_N_S_test_S_temperature_sensor_P_pinctrl_MY_INDEX_LEN);
 printk("len with api: %d\n", 2);
}
```

Because manual use of `##` in `FOO` misses that you actually have to recursively expand again to get `MY_INDEX` to resolve to an integer literal.

File: include/devicetree/pinctrl.h
Comment (mbolivar-nordic) [reply include/devicetree/pinctrl.h:153]: > same comment as above.

resolving since it's the same response

File: dts/bindings/test/vnd,adc-temp-sensor.yaml
Comment (mbolivar-nordic) [reply dts/bindings/test/vnd,adc-temp-sensor.yaml:29]: > It's part of #37572, feel free to take that file.

Thanks but no thanks. Pulling that file in without updating all the in-tree bindings to include it feels wrong to me, and that's outside the scope of this PR.

File: soc/arm/nuvoton_npcx/common/soc_dt.h
Comment (mbolivar-nordic) [reply soc/arm/nuvoton_npcx/common/soc_dt.h]: Thanks @ChiHuaL; now fixed.

**Review by galak — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
Last push is just documentation fixes.

**Comment by mbolivar-nordic** [comment]
I'm adding a DNM while we're discussing the layers that will be built on top of this.

There are currently two draft proposals:

- https://github.com/zephyrproject-rtos/zephyr/pull/37621 (by @galak)
- https://github.com/zephyrproject-rtos/zephyr/pull/37572 (by @gmarull)

In either one of these, some of the macros introduced here and used in soc-specific headers may become redundant.

**Comment by mbolivar-nordic** [comment]
Latest changes:

- add new macros (requested by @gmarull):
-- DT_PINCTRL_IDX_TO_NAME_TOKEN(node_id, pc_idx)
-- DT_PINCTRL_IDX_TO_NAME_UPPER_TOKEN(node_id, pc_idx)
-- DT_PINCTRL_IDX_TO_NAME_TOKEN()
-- DT_PINCTRL_IDX_TO_NAME_UPPER_TOKEN()
- make existing macros consume lowercase-and-underscores names tokens
- update tests
- add missing tests

File: include/devicetree/pinctrl.h
Comment (gmarull) [reply include/devicetree/pinctrl.h:189]: question: `DT_NUM_*` macros do not follow the `DT_PINCTRL_*` namespace convention, not sure if important.

Comment (mbolivar-nordic) [reply include/devicetree/pinctrl.h:189]: > question: `DT_NUM_*` macros do not follow the `DT_PINCTRL_*` namespace convention, not sure if important.

https://docs.zephyrproject.org/latest/reference/devicetree/api.html#c.DT_NUM_REGS
https://docs.zephyrproject.org/latest/reference/devicetree/api.html#c.DT_NUM_IRQS

Yep, done on purpose for consistency with other APIs like the above.

**Review by brockus-zephyr — APPROVED** [review]
> Reviewed 2 of 3 files at r1, 18 of 18 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/37644)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @ABOSTM, @FRASTM, @henrikbrixandersen, @ioannisg, @jackrosenthal, @katsuster, @keith-zephyr, @kgugala, @MaureenHelm, @mgielda, @nandojve, @nashif, @pgielda, and @sjg20)

**Comment by mbolivar-nordic** [comment]
Rebased and removed DNM. I'm going to leave it up to @galak to decide if we want this in for LTS>

**Comment by cfriedt** [comment]
@galak - feel free to merge or if you would like to hold off apply the DNM label again

**Review by nandojve — APPROVED** [review]

**Review by gmarull — APPROVED** [review]

**Review by brockus-zephyr — APPROVED** [review]
> Reviewed 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/37644)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @ABOSTM, @FRASTM, @henrikbrixandersen, @ioannisg, @jackrosenthal, @katsuster, @keith-zephyr, @kgugala, @MaureenHelm, @mgielda, @nashif, @pgielda, and @sjg20)

**Pull request merged**
